File: python/linkedlist/ReverseNodesinkGroup.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        if head is None or head.next is None:
            return head
            
        count_head = head
        counter = 0
        while count_head:
            counter += 1
            count_head = count_head.next
        
        if counter < k:
            return head
        
        num_groups = counter / k
        
        dummy = ListNode(0)
        dummy.next = head
        start = dummy
        curr = dummy.next
        for i in xrange(num_groups):
            for j in xrange(k-1):
                t0 = start.next
                t1 = curr.next.next
                start.next = curr.next
                curr.next.next = t0
                curr.next = t1
            start = curr
            curr = curr.next
                
        return dummy.next
                
if __name__ == "__main__":
    solution = Solution()
    l0 = ListNode(1)
    l1 = ListNode(2)
    l2 = ListNode(3)
    l0.next = l1
    l1.next = l2

    p = solution.reverseKGroup(l0, 2)
    while p:
        print p.val
        p = p.next

File: java/ReverseNodesInKGroup.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverse (ListNode p, int k) {
        int j = 0;
        ListNode new_tail = p.next;
        while (j < k - 1) {
            ListNode tmp = new_tail.next;
            new_tail.next = new_tail.next.next;
            tmp.next = p.next;
            p.next = tmp;
            j += 1;
        }
        return new_tail;
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        Boolean reverse = true;
        ListNode p = dummy;
        while (true) {
            int i = 0;
            ListNode tmp = p;
            while (i < k) {
                if (tmp.next == null) {
                    reverse = false;
                    break;
                }
                tmp = tmp.next;
                i += 1;
            }
            if (reverse == false) {
                break;
            }
            p = this.reverse(p, k);
        }
        return dummy.next;
    }
}
